Fixed defects with script

StuWill opened the pull request:
### What 
Fixing bugs with resource 

#### Last commit 
Fixed defects with script

Updated concourse resource type reference links

Added in checks for if variables aren't supplied, and added in defaults
that could fail if not supplied

Accounted for if a path of '/' is given

Added shell options, including to exit on error rather than run with errors

### Why
There were some bugs with this resource that meant it wouldn't exit if there were errors and would continue failing through the scripts. It was also pointed out that it is bad practice not to have checks for user input (that didn't have default's set).

### How


Check diffs and merge, no major functionality changes.

StuWill committed:
Fixed defects with script

Updated concourse resource type reference links

Added in checks for if variables aren't supplied, and added in defaults
that could fail if not supplied

Accounted for if a path of '/' is given

Added shell options, including to exit on error rather than run with errors

combor commented on assets/out:
```diff
@@ -18,27 +18,43 @@ fi
-AWS_SECRET_ACCESS_KEY=$(echo "${payload}" | jq -r '.source.secret_access_key')
+AWS_ACCESS_KEY_ID=$(echo "$payload" | jq -r '.source.access_key_id')
+if [ -z "${AWS_ACCESS_KEY_ID}" ]; then
+  echo "No access_key_id supplied, needed if accessing minio."
```
This needs to exit 1 if there is no key id supplied. Same for access key

combor commented on assets/in:
```diff
@@ -16,26 +18,42 @@ fi
-AWS_SECRET_ACCESS_KEY=$(echo "$payload" | jq -r '.source.secret_access_key')
+AWS_ACCESS_KEY_ID=$(echo "${payload}" | jq -r '.source.access_key_id')
+if [ -z "${AWS_ACCESS_KEY_ID}" ]; then
+  echo "No access_key_id supplied, needed if accessing minio"
```
This needs to exit 1 if there is no key id supplied. Same for access key

combor commented on assets/check:
```diff
@@ -1,27 +1,43 @@
 # export for `aws` cli
 AWS_ACCESS_KEY_ID=$(echo "$payload" | jq -r '.source.access_key_id')
+if [ -z "${AWS_ACCESS_KEY_ID}" ]; then
+  echo "No access_key_id supplied, needed if accessing minio."
```
This needs to exit 1 if there is no key id supplied. Same for access key

combor commented on assets/check:
```diff
@@ -1,27 +1,43 @@
+endpoint_url="$(echo "${payload}" | jq -r '.source.endpoint // ""')"
+region="$(echo "${payload}" | jq -r '.source.region // "us-east-1"')"
+use_v4="$(echo "${payload}" | jq -r '.source.use_v4 // ""')"
+path="$(echo "${payload}" | jq -r '.source.path // ""')"
```
if you provide a random string path, the version is still produced which is wrong. There should be an error that such path doesnt exist.

combor commented on assets/in:
```diff
@@ -16,26 +18,42 @@ fi
 use_v4="$(echo "${payload}" | jq -r '.source.use_v4 // ""')"
-path=$(echo "${payload}" | jq -r '.source.path // ""')
-suffix=$(echo "${payload}" | jq -r '.source.suffix // ""')
+path="$(echo "${payload}" | jq -r '.source.path // ""')"
```
If you provide a random string path that doesn't exist the script finishes successfully. IMHO it should fail

combor reviewed (changes requested):
I left some comments.

Tested with json:
```
{
  "source": {
    "endpoint": "http://lab-minio.you-know-what.com:9001",
    "access_key_id": "foo",
    "secret_access_key": "bar",
    "bucket": "dev6",
    "path": "nonexistent",
    "options": []
  }
}
```
cat /tmp/x | ./out /tmp/test/
cat /tmp/x | ./in /tmp/test/
cat /tmp/x | ./check /tmp/test/

StuWill committed:
Fixes for pr comments

Have added checks for if s3 provider endpoint is minio then exit on no
creds being supplied. Readme suggests there is functionality with ec2
where creds aren't necessarily needed, so don't want to exit for
everything.

Have added a check for the path variable, will try to ls the bucket/path
and if the path does not exist it will error and will exit script.

combor commented on assets/out:
```diff
@@ -68,10 +90,14 @@ if [ ! "${encryption_key}" = "" ];then
 
+
+echo "Checking path in S3..."
+aws s3 ls "s3://${bucket}/${path}" --endpoint "${endpoint_url}" --no-verify-ssl > /dev/null || echo "Path check exited with code [$?], check path provided exists" && exit 1
```
we have `endpoint_opt="--endpoint-url=${endpoint_url}"` earlier, it would be good to use it.
Also this is missing `${ssl_verification_opt}` but instead hardcoding it which is not correct.

Please check for this in all occurances of this command.

combor commented on assets/in:
```diff
@@ -16,26 +18,49 @@ fi
 encryption_key="$(echo "${payload}" | jq -r '.source.encryption_key // ""')"
 skip_ssl_verification="$(echo "$payload" | jq -r '.source.skip_ssl_verification // ""')"
 ca_bundle="$(echo "$payload" | jq -r '.source.ca_bundle // ""')"
+minio_identifier="minio" # used to check if endpoint_url contains 'minio'
```
Does it have to be `minio`? This is non FIL specific resource and we can't assume this.

combor commented on assets/in:
```diff
@@ -16,26 +18,49 @@ fi
+fi
 AWS_SECRET_ACCESS_KEY=$(echo "$payload" | jq -r '.source.secret_access_key')
+if [ -z "${AWS_SECRET_ACCESS_KEY}" ] || [ "${AWS_SECRET_ACCESS_KEY}" = "null" ]; then
+  if [ "${endpoint_url#*$minio_identifier}" != "${endpoint_url}" ]; then
```
I wouldnt differentiate beteen minio and aws. It doesn't matter as both are an implemetation of s3 api. We do `aws s3 ls ` below to check if path exists. This is enough to figure out if correct credentials were supplied.

combor reviewed (changes requested):
More comments :)

combor commented:
Shal we remove "DO NOT MERGE" - from the desc? It scares me a bit ;)

StuWill committed:
Avoiding hardcoding check for specific s3 provider

Aws ls command works for checking both path and creds, so removed prior
cred checks.
